### PR TITLE
[x/programs] Improve runtime Config defaults 

### PR DESCRIPTION
**Pull request by hexfusion**
This PR adds.

- improved `Config` default handling
- support for serializing the runtime `ConfigBuilder`
- allow multiple instances of the runtime.Config to be passed to additional runtimes. Currently this would fail as the wasmtime.Config can only be used once per engine. This is important so we can ensure program to program calls use the same runtime configuration as the parent.
- bumps defaultLimitMaxMemory from 16 to 18 pages. We found that 18 pages is required for all of our current examples.
- Add testing only mode which allows for non production debug modes by enabling `Wasi` preview for the upcoming simulator work.
- Removed `maxFee` from `ConfigBuilder` and added to Initialize()
- renamed `TransferUnits` -> `TransferUnitsTo`

### CODE REVIEW

**Review by exdx — DISMISSED** [review]
lgtm

File: x/programs/examples/token_test.go [hunk 1]
```diff
@@ -25,7 +25,7 @@ var (
 	log = logging.NewLogger(
 		"",
 		logging.NewWrappedCore(
-			logging.Info,
+			logging.Debug,
```
Comment (hexfusion) [inline x/programs/examples/token_test.go:28]: nit

File: x/programs/rust/scripts/build.sh [hunk 2]
```diff
@@ -1,15 +1,30 @@
...
-  --target wasm32-unknown-unknown \
-  --target-dir "$CARGO_TARGET_DIR" \
+  --target "${target}" \
+  --target-dir "${target_dir}" \
   --release
+
+# Copy the compiled program into the build directory
+cp ./"${target_dir}"/"${target}"/release/*.wasm "./${target_dir}/"
+
+echo "Wasm files are in ./${target_dir}"
```
Comment (hexfusion) [inline x/programs/rust/scripts/build.sh:30]: nit

File: x/programs/runtime/config.go
Comment (patrick-ogrady) [inline x/programs/runtime/config.go:1]: Much cleaner. Nice.

File: x/programs/runtime/runtime.go [hunk 3]
```diff
@@ -82,13 +87,30 @@ func (r *WasmRuntime) Initialize(ctx context.Context, programBytes []byte) (err
...
 	// setup metering
 	r.meter = NewMeter(r.store)
 	_, err = r.meter.AddUnits(r.cfg.meterMaxUnits)
 	if err != nil {
 		return err
 	}
 
+	// reset the config max units to 0 this ensures that any subsequent usage of
+	// this config will not inherit the meter units. ex (program -> program)
+	r.cfg.meterMaxUnits = 0
```
Comment (patrick-ogrady) [inline x/programs/runtime/runtime.go:112]: This must've been a fun bug.

Comment (patrick-ogrady) [reply x/programs/runtime/runtime.go:112]: Should it instead be set to the remaining units? Or is that handled elsewhere?

File: x/programs/rust/README.md [hunk 4]
```diff
@@ -18,6 +18,17 @@ cargo build --target wasm32-unknown-unknown --target-dir $CARGO_TARGET_DIR --rel
 ./scripts/build.sh
 ```
 
+## Debugging
+
+While developing programs you can optionally compile the program with wasi
+support. This provides for example logging from your program. The compile Wasm
```
Comment (patrick-ogrady) [inline x/programs/rust/README.md:24]: If you leave the logs in and don't use `TESTING_ONLY`, what happens? If you have to remove them, I'd make sure to say that here as a reminder.

File: x/programs/runtime/runtime.go (hunk 3 above)
Comment (hexfusion) [reply x/programs/runtime/runtime.go:112]: Good question, the idea is the caller sets the maxfee for the entire execution. once the parent runtime sets the meter I want this cfg value to be 0. This way if we use the config again for program to program the expectation is the meter should start at 0. We then can handle remaining units with `TransferUnits`. You can find an example of this in the program import example[1].

 https://github.com/ava-labs/hypersdk/blob/9c184578d8d1536964abe34055754e15bde644fe/x/programs/runtime/meter.go#L55
 
 [1] https://github.com/ava-labs/hypersdk/blob/9c184578d8d1536964abe34055754e15bde644fe/x/programs/examples/imports/program/program.go#L170

File: x/programs/rust/README.md (hunk 4 above)
Comment (hexfusion) [reply x/programs/rust/README.md]: improved the docs on this.

File: x/programs/rust/README.md [hunk 5]
```diff
@@ -20,10 +20,14 @@ cargo build --target wasm32-unknown-unknown --target-dir $CARGO_TARGET_DIR --rel
...
-While developing programs you can optionally compile the program with wasi
-support. This provides for example logging from your program. The compile Wasm
-will be automatically supported by the VM simulator or if you explicitly set
-WithEnableTestingOnlyMode(true) for the runtime `Config`.
+While developing programs you can optionally compile the program with `Wasi`
+support. This allows, for example, logging from your program. The compiled
+`Wasm` will be automatically supported by the VM simulator, or if you explicitly
+set `WithEnableTestingOnlyMode` for the runtime `Config`. 
+
+**NOTE**: Once testing is complete, remove all print statements from your
```
Comment (patrick-ogrady) [inline x/programs/rust/README.md:28]: If you keep them around, will it error or just skip it?

File: x/programs/runtime/runtime.go (hunk 3 above)
Comment (patrick-ogrady) [reply x/programs/runtime/runtime.go:112]: Ah ok. We should probably document this very explicitly (maybe right there?)

File: x/programs/rust/README.md (hunk 5 above)
Comment (hexfusion) [reply x/programs/rust/README.md:28]: there is no error if you recompile, if you don't it will error.

File: x/programs/runtime/runtime.go (hunk 3 above)
Comment (hexfusion) [reply x/programs/runtime/runtime.go:112]: Made a few other changes here. I think its important that there is only one flow for populating the meter with a single config to reduce the scope of leaky metering.

- new cfg populated by genesis sets `maxFee`
- any other inline runtimes must `TransferUnits` meter <-> meter.

Comment (hexfusion) [reply x/programs/runtime/runtime.go]: I thought this over and the Unit fee really should not be part of the cfg for the reasons I am trying to escape :). I have been back and forth on this but I think it should be a param to Intialize.

File: x/programs/rust/README.md (hunk 5 above)
Comment (hexfusion) [reply x/programs/rust/README.md]: Decided to leave the script and clarify the docs

File: x/programs/examples/counter.go [hunk 6]
```diff
@@ -1,197 +0,0 @@
-// Copyright (C) 2023, Ava Labs, Inc. All rights reserved.
-// See the file LICENSE for licensing terms.
-
-package examples
```
Comment (hexfusion) [inline x/programs/examples/counter.go:4]: these are really just tests

File: x/programs/runtime/runtime.go (hunk 3 above)
Comment (patrick-ogrady) [reply x/programs/runtime/runtime.go]: Ok sounds good.

**Review by exdx — DISMISSED** [review]

File: x/programs/rust/scripts/build.sh [hunk 7]
```diff
@@ -1,15 +1,30 @@
...
+# $ TESTING_ONLY=1 ./scripts/build.sh .
+#
+# If TESTING_ONLY=1 is passed we compile the program with wasi support for
+# easier debugging.
 set -x
 
-# Set a default value for CARGO_TARGET_DIR if it's not provided as an argument
-CARGO_TARGET_DIR="${1:-../examples/testdata/target}"
+target="wasm32-unknown-unknown"
+if [ -n "${TESTING_ONLY:-}" ]; then
```
Comment (patrick-ogrady) [inline x/programs/rust/scripts/build.sh:10]: Can we just call this `DEBUG`?

File: x/programs/runtime/meter.go [hunk 8]
```diff
@@ -52,8 +52,9 @@ func (m *meter) AddUnits(units uint64) (uint64, error) {
 	return m.GetBalance(), nil
 }
 
-func (m *meter) TransferUnits(to Meter, units uint64) (uint64, error) {
-	// TODO: add rollback support
+// TransferUnitsTo moves units from this meter to another meter.
+func (m *meter) TransferUnitsTo(to Meter, units uint64) (uint64, error) {
+	// TODO: add rollback support?
```
[the remark this answers is not in the document]
Comment (hexfusion) [reply x/programs/runtime/meter.go:57]: true only concern was if the transfer itself failed but not much that really can be done. will drop

**Pull request merged**